network/dnstruncate/packet_proxy.go

```diff
@@ -0,0 +1,157 @@
+// without sending the requests to the actual DNS resolver. It sets the TC (truncated) bit in the DNS response header
+// to tell the caller to resend the DNS request over TCP.
+//
+// Multiple goroutines may invoke methods on a dnsTruncateProxy simultaneously.
+type dnsTruncateRequestHandler struct {
+	closed     atomic.Bool
+	respWriter network.PacketResponseReceiver
+}
+
+// Compilation guard against interface implementation
+var _ network.PacketProxy = (*dnsTruncateProxy)(nil)
+var _ network.PacketRequestSender = (*dnsTruncateRequestHandler)(nil)
+
+// NewPacketProxy creates a new [network.PacketProxy] that can be used to handle DNS requests if the remote proxy
+// doesn't support UDP traffic. It sets the TC (truncated) bit in the DNS response header to tell the caller to resend
+// the DNS request over TCP.
+//
+// This [network.PacketProxy] should only be used if the remote proxy server doesn't support UDP traffic at all. Note
+// that all other non-DNS UDP packets will be dropped by this [network.PacketProxy].
+func NewPacketProxy() (network.PacketProxy, error) {
+	return &dnsTruncateProxy{}, nil
+}
+
+// NewSession implements [network.PacketProxy].NewSession(). It creates a new [network.PacketRequestSender] that will
+// set the TC (truncated) bit and write the response to `respWriter`.
+func (p *dnsTruncateProxy) NewSession(respWriter network.PacketResponseReceiver) (network.PacketRequestSender, error) {
+	if respWriter == nil {
+		return nil, errors.New("respWriter is required")
+	}
+	return &dnsTruncateRequestHandler{
+		respWriter: respWriter,
+	}, nil
+}
+
+// Close implements [network.PacketRequestSender].Close(), and it closes the corresponding
+// [network.PacketResponseReceiver].
+func (h *dnsTruncateRequestHandler) Close() error {
+	if !h.closed.CompareAndSwap(false, true) {
+		return network.ErrClosed
+	}
+	h.respWriter.Close() // TODO(junyi): potential inf loop (rw.Close -> req.Close -> rw.Close ...)
+	return nil
+}
+
+// WriteTo implements [network.PacketRequestSender].WriteTo(). It parses a packet from p, and determines whether it is
+// a valid DNS request. If so, it will write the DNS response with TC (truncated) bit set to the corresponding
+// [network.PacketResponseReceiver] passed to NewSession. If it is not a valid DNS request, the packet will be
+// discarded and returns an error.
+func (h *dnsTruncateRequestHandler) WriteTo(p []byte, destination net.Addr) (int, error) {
+	if h.closed.Load() {
+		return 0, network.ErrClosed
+	}
+	resolverAddr, err := net.ResolveUDPAddr("udp", destination.String())
+	if err != nil {
+		return 0, fmt.Errorf("non-UDP %w, the DNS resolver is not a valid UDP address: %w", network.ErrUnsupported, err)
+	}
+	if resolverAddr.Port != dnsServerPort {
+		return 0, fmt.Errorf("non-DNS UDP %w, target server's port is %v rather than %v",
+			network.ErrUnsupported, resolverAddr.Port, dnsServerPort)
+	}
```

Comment:
Perhaps create a port unreachable error, so the caller can generate a ICMP packet in the future?

```suggestion
	if resolverAddr.Port != dnsServerPort {
		return 0, fmt.Errorf("non-DNS UDP is not supported: %w", network.ErrPortUnreachable(destAddr.Port))
	}
```